### DISCUSSION TITLE
Replace CoinGecko with decentralized rate providers queried over Tor similar to how Bisq does

### DISCUSSION THREAD

**Discussion by wiz** [btcpayserver/btcpayserver#1860] [Ideas / Feature Requests]
Currently BTCPay has a TTP / CPOF using CoinGecko as its primary rate provider. Previously Bitcoin Average shut off its free API service and BTCPay users were affected. Additionally, all these queries are currently performed over clearnet, so user IP addresses are exposed and logged by data providers. This presents a privacy and censorship risk to BTCPay.

Bisq recently addressed this issue of a centralized rate provider by replacing Bitcoin Average with its own implementation of a price service node, which queries 35+ exchanges. The full list is available at https://bisq.wiki/Bisq_Price_Indices

The 5 pricenodes that Bisq uses, and the people operating them, are:

wizpriceje6q5tdrxkyiazsgu7irquiqjy2dptezqhrtu7l2qelqktid (@wiz)
emzypricpidesmyqg2hc6dkwitqzaxrqnpkdg3ae2wef5znncu2ambqd (@Emzy)
devinpndvdwll4wiqcyq5e7itezmarg7rzicrvf6brzkwxdm374kmmyd (@devinbileck)
aprcndeiwdrkbf4fq7iozxbd27dl72oeo76n7zmjwdi4z34agdrnheyd (@mrosseel)
ro7nv73awqs3ga2qtqeqawrjpbxwarsazznszvr6whv7tes5ehffopid (@alexej996)

Example query: http://wizpriceje6q5tdrxkyiazsgu7irquiqjy2dptezqhrtu7l2qelqktid.onion/getAllMarketPrices

Trusted BTCPay contributors can also easily setup their own instances of the Bisq Pricenode, The source for the Bisq Pricenode is available at https://github.com/bisq-network/bisq/tree/master/pricenode

Rough implementation plan:

1) Add privoxy as a new Docker image to btcpayserver-docker, bound to localhost:8118, connects to Tor on localhost:9050

Found a reference Dockerfile for privoxy here https://github.com/vimagick/dockerfiles/blob/master/privoxy/Dockerfile

2) Query the Bisq Pricenodes onions from .NET using Nicolas's hacked HTTP proxy through privoxy on localhost:8118

Nicolas just needs to `GET http://foo.onion/getAllMarketPrices` from his code

3) Add the list of decentralized rate providers as a randomized list into BTCPay

**Comment by Kukks** [comment]
We already have a built-in tor socks5 proxy HTTP client in BTCPay so we don't need privoxy afaik.
Should be easy enough to integrate the rate provider in BTCPay but unsure what the best way to deal with these being community-run. I guess there needs to be a config to load the price node uris.

**Comment by wiz** [comment]
Okay cool, yeah if you can already query Tor HS onions then no need for any dev work I guess, just configuration... can you try querying the onions I posted above as a quick test?

**Reply by nixbitcoin** [reply]
Can we just add a rate provider's Tor HS onion to the default rate providers?

**Comment by wiz** [comment]
@NicolasDorier alternatively, I could make a PR to add the Bisq Pricenode into btcpayserver-docker as an optional component, so btcpay could just query its local docker image for prices?
